File: skills/rust-release-orchestrator/scripts/orchestrate.py
#!/usr/bin/env python3
"""
Autonomous Rust workspace release orchestrator.

Usage:
  python orchestrate.py [--dry-run] [--resume] [--workspace <path>] [--skip-gates]

State file: .release-state.json in the workspace root.
Resume: re-run with --resume; already-published crates are skipped.
"""

import argparse
import json
import os
import subprocess
import sys
import time
from pathlib import Path

# ── ANSI colours ──────────────────────────────────────────────────────────────
GREEN = "\033[92m"
YELLOW = "\033[93m"
RED = "\033[91m"
CYAN = "\033[96m"
BOLD = "\033[1m"
RESET = "\033[0m"


def info(msg):
    print(f"{CYAN}[INFO]{RESET}  {msg}")


def ok(msg):
    print(f"{GREEN}[OK]{RESET}    {msg}")


def warn(msg):
    print(f"{YELLOW}[WARN]{RESET}  {msg}")


def fail(msg):
    print(f"{RED}[FAIL]{RESET}  {msg}", file=sys.stderr)


def step(msg):
    print(f"\n{BOLD}{CYAN}══ {msg} ══{RESET}")


# ── Shell helpers ──────────────────────────────────────────────────────────────


def run(cmd, cwd=None, capture=False, check=True):
    """Run a shell command; return (stdout, returncode)."""
    result = subprocess.run(
        cmd,
        shell=True,
        cwd=cwd,
        stdout=subprocess.PIPE if capture else None,
        stderr=subprocess.PIPE if capture else None,
        text=True,
    )
    if check and result.returncode != 0:
        stdout = result.stdout or ""
        stderr = result.stderr or ""
        raise RuntimeError(f"Command failed: {cmd}\n{stdout}\n{stderr}")
    return (result.stdout or ""), result.returncode


# ── Dependency graph ───────────────────────────────────────────────────────────


def compute_publish_order(workspace_root: Path) -> list[dict]:
    """
    Parse `cargo metadata` and return publishable crates in topological order
    (dependencies before dependents).
    """
    raw, _ = run(
        "cargo metadata --no-deps --format-version 1", cwd=workspace_root, capture=True
    )
    meta = json.loads(raw)

    # Map name -> package info for workspace members
    members = {
        p["name"]: p for p in meta["packages"] if p["id"] in meta["workspace_members"]
    }
    publishable = {
        name: pkg for name, pkg in members.items() if pkg.get("publish") != []
    }

    # Build adjacency: dep_name -> set of pkgs that depend on it
    rdeps: dict[str, set] = {n: set() for n in publishable}
    for name, pkg in publishable.items():
        for dep in pkg.get("dependencies", []):
            dep_name = dep["name"]
            if dep_name in publishable:
                rdeps[dep_name].add(name)

    # Kahn's algorithm for topological sort
    in_degree = {n: 0 for n in publishable}
    for name in publishable:
        for dependent in rdeps[name]:
            in_degree[dependent] += 1

    queue = [n for n, d in in_degree.items() if d == 0]
    order = []
    while queue:
        queue.sort()  # deterministic
        n = queue.pop(0)
        order.append(publishable[n])
        for dependent in sorted(rdeps[n]):
            in_degree[dependent] -= 1
            if in_degree[dependent] == 0:
                queue.append(dependent)

    if len(order) != len(publishable):
        raise RuntimeError("Cycle detected in workspace dependency graph!")

    return order


# ── Quality gates ──────────────────────────────────────────────────────────────


def run_gates(workspace_root: Path, dry_run: bool) -> bool:
    """Run fmt / clippy / nextest. Auto-fix what can be fixed. Return True on success."""
    step("Quality gates")

    # fmt
    info("cargo fmt --all --check …")
    _, rc = run(
        "cargo fmt --all --check", cwd=workspace_root, capture=True, check=False
    )
    if rc != 0:
        warn("fmt check failed — auto-formatting …")
        run("cargo fmt --all", cwd=workspace_root)
        ok("Auto-formatted")
    else:
        ok("fmt clean")

    # clippy
    info("cargo clippy --all-targets …")
    out, rc = run(
        "cargo clippy --all-targets -- -D warnings 2>&1",
        cwd=workspace_root,
        capture=True,
        check=False,
    )
    if rc != 0:
        warn("Clippy warnings found — attempting auto-fix …")
        _, fix_rc = run(
            "cargo clippy --fix --allow-dirty --allow-staged --all-targets -- -D warnings 2>&1",
            cwd=workspace_root,
            capture=True,
            check=False,
        )
        # Re-check
        _, rc2 = run(
            "cargo clippy --all-targets -- -D warnings 2>&1",
            cwd=workspace_root,
            capture=True,
            check=False,
        )
        if rc2 != 0:
            fail("Clippy still failing after auto-fix. Fix manually and re-run.")
            return False
        ok("Clippy clean after auto-fix")
    else:
        ok("Clippy clean")

    # tests
    info("cargo nextest run …")
    _, rc = run("cargo nextest run 2>&1", cwd=workspace_root, capture=True, check=False)
    if rc != 0:
        fail("Tests failed. Fix and re-run.")
        return False
    ok("All tests passed")

    return True


# ── Publish with retry ─────────────────────────────────────────────────────────

ALREADY_PUBLISHED_MARKERS = [
    "already exists",
    "crate version `",  # "crate version `X.Y.Z` is already uploaded"
    "already uploaded",
]
RATE_LIMIT_MARKERS = [
    "too many requests",
    "rate limit",
    "429",
]


def _is_already_published(output: str) -> bool:
    lo = output.lower()
    return any(m in lo for m in ALREADY_PUBLISHED_MARKERS)


def _is_rate_limited(output: str) -> bool:
    lo = output.lower()
    return any(m in lo for m in RATE_LIMIT_MARKERS)


def publish_crate(
    pkg: dict, workspace_root: Path, dry_run: bool, max_retries: int = 5
) -> bool:
    """Publish one crate with exponential-backoff retry. Return True on success."""
    name = pkg["name"]
    version = pkg["version"]
    crate_path = Path(pkg["manifest_path"]).parent

    flag = "--dry-run" if dry_run else ""
    cmd = f"cargo publish {flag} 2>&1"

    delay = 5  # seconds
    for attempt in range(1, max_retries + 1):
        info(f"Publishing {name} v{version} (attempt {attempt}/{max_retries}) …")
        out, rc = run(cmd, cwd=crate_path, capture=True, check=False)

        if rc == 0:
            ok(f"{name} v{version} published {'(dry-run)' if dry_run else ''}")
            return True

        if _is_already_published(out):
            ok(f"{name} v{version} already published — treating as success")
            return True

        if _is_rate_limited(out):
            wait = delay * (2 ** (attempt - 1))
            warn(f"Rate limited. Waiting {wait}s before retry …")
            time.sleep(wait)
            continue

        # Hard failure
        fail(f"{name} publish failed:\n{out}")
        if attempt < max_retries:
            wait = delay * (2 ** (attempt - 1))
            warn(f"Retrying in {wait}s …")
            time.sleep(wait)
        else:
            return False

    return False


# ── State management ───────────────────────────────────────────────────────────


def load_state(state_path: Path) -> dict:
    if state_path.exists():
        return json.loads(state_path.read_text())
    return {"published": [], "gates_passed": False}


def save_state(state_path: Path, state: dict):
    state_path.write_text(json.dumps(state, indent=2))


def clear_state(state_path: Path):
    if state_path.exists():
        state_path.unlink()


# ── Report ─────────────────────────────────────────────────────────────────────


def emit_report(order: list[dict], results: dict, dry_run: bool, elapsed: float):
    width = 60
    print("\n" + "═" * width)
    print(f"  {'DRY-RUN ' if dry_run else ''}RELEASE REPORT")
    print("═" * width)
    published = [r for r in results.values() if r == "published"]
    skipped = [r for r in results.values() if r == "skipped"]
    failed = [r for r in results.values() if r == "failed"]

    for pkg in order:
        name = pkg["name"]
        ver = pkg["version"]
        status = results.get(name, "?")
        icon = {"published": "✓", "skipped": "↩", "failed": "✗"}.get(status, "?")
        colour = {"published": GREEN, "skipped": YELLOW, "failed": RED}.get(
            status, RESET
        )
        print(f"  {colour}{icon}{RESET}  {name:<30} v{ver}  [{status}]")

    print("─" * width)
    print(f"  Published : {len(published)}")
    print(f"  Skipped   : {len(skipped)}  (already on crates.io)")
    print(f"  Failed    : {len(failed)}")
    print(f"  Elapsed   : {elapsed:.1f}s")
    if dry_run:
        print(
            f"\n  {YELLOW}This was a DRY RUN — nothing was actually published.{RESET}"
        )
        print(f"  Re-run without --dry-run to publish for real.")
    print("═" * width + "\n")
    return len(failed) == 0


# ── Main ───────────────────────────────────────────────────────────────────────


def main():
    parser = argparse.ArgumentParser(description="Rust workspace release orchestrator")
    parser.add_argument(
        "--dry-run", action="store_true", help="Skip actual cargo publish"
    )
    parser.add_argument(
        "--resume", action="store_true", help="Resume from previous state"
    )
    parser.add_argument("--workspace", default=".", help="Path to workspace root")
    parser.add_argument(
        "--skip-gates", action="store_true", help="Skip fmt/clippy/test"
    )
    args = parser.parse_args()

    workspace = Path(args.workspace).resolve()
    state_path = workspace / ".release-state.json"
    t0 = time.time()

    step("Rust Workspace Release Orchestrator")
    info(f"Workspace : {workspace}")
    info(f"Dry-run   : {args.dry_run}")
    info(f"Resume    : {args.resume}")

    # State
    state = (
        load_state(state_path)
        if args.resume
        else {"published": [], "gates_passed": False}
    )

    # 1. Compute publish order
    step("Computing publish order")
    order = compute_publish_order(workspace)
    info(f"Publish order ({len(order)} crates):")
    for i, pkg in enumerate(order, 1):
        print(f"  {i}. {pkg['name']} v{pkg['version']}")

    # 2. Quality gates (skip if already passed in this session and resuming)
    if not args.skip_gates and not state.get("gates_passed"):
        if not run_gates(workspace, args.dry_run):
            fail(
                "Quality gates failed. Fix issues and re-run (add --resume to continue from here)."
            )
            state["gates_passed"] = False
            save_state(state_path, state)
            sys.exit(1)
        state["gates_passed"] = True
        save_state(state_path, state)
    elif state.get("gates_passed"):
        ok("Quality gates already passed (resuming) — skipping")

    # 3. Publish each crate in order
    step("Publishing crates")
    results = {}
    for pkg in order:
        name = pkg["name"]
        if name in state["published"]:
            ok(f"{name} already published in previous run — skipping")
            results[name] = "skipped"
            continue

        success = publish_crate(pkg, workspace, args.dry_run)
        if success:
            results[name] = "published"
            if not args.dry_run:
                state["published"].append(name)
                save_state(state_path, state)
        else:
            results[name] = "failed"
            fail(f"{name} failed. State saved. Re-run with --resume to continue.")
            save_state(state_path, state)
            # Emit partial report and exit
            emit_report(order, results, args.dry_run, time.time() - t0)
            sys.exit(1)

    # 4. Report
    elapsed = time.time() - t0
    success = emit_report(order, results, args.dry_run, elapsed)

    # Clean up state on full success
    if success and not args.dry_run:
        clear_state(state_path)
        info("State file cleaned up")

    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: skills/rust-release-orchestrator/scripts/orchestrate.rs
#!/usr/bin/env rust-script
//! Rust workspace release orchestrator.
//!
//! ```cargo
//! [dependencies]
//! anyhow = "1"
//! clap = { version = "4", features = ["derive"] }
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! ```

use anyhow::{bail, Context, Result};
use clap::Parser;
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet, VecDeque};
use std::path::{Path, PathBuf};
use std::process::{Command, Output};
use std::thread;
use std::time::{Duration, Instant};

#[derive(Debug, Parser)]
#[command(about = "Publish a Rust workspace in dependency order")]
struct Args {
    #[arg(long)]
    dry_run: bool,
    #[arg(long)]
    resume: bool,
    #[arg(long, default_value = ".")]
    workspace: PathBuf,
    #[arg(long)]
    skip_gates: bool,
}

#[derive(Debug, Deserialize)]
struct Metadata {
    packages: Vec<Package>,
    workspace_members: Vec<String>,
}

#[derive(Debug, Clone, Deserialize)]
struct Package {
    id: String,
    name: String,
    version: String,
    manifest_path: PathBuf,
    publish: Option<Vec<String>>,
    #[serde(default)]
    dependencies: Vec<Dependency>,
}

#[derive(Debug, Clone, Deserialize)]
struct Dependency {
    name: String,
}

#[derive(Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
struct ReleaseState {
    published: Vec<String>,
    gates_passed: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum ResultStatus {
    Published,
    Skipped,
    Failed,
}

fn info(message: impl std::fmt::Display) {
    println!("[INFO]  {message}");
}
fn ok(message: impl std::fmt::Display) {
    println!("[OK]    {message}");
}
fn warn(message: impl std::fmt::Display) {
    println!("[WARN]  {message}");
}
fn fail(message: impl std::fmt::Display) {
    eprintln!("[FAIL]  {message}");
}
fn step(message: impl std::fmt::Display) {
    println!("\n══ {message} ══");
}

fn run(cwd: &Path, program: &str, args: &[&str]) -> Result<Output> {
    Command::new(program)
        .args(args)
        .current_dir(cwd)
        .output()
        .with_context(|| format!("running {program} {}", args.join(" ")))
}

fn require_success(output: Output, description: &str) -> Result<Output> {
    if output.status.success() {
        return Ok(output);
    }
    bail!(
        "{description} failed:\n{}{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    )
}

fn compute_publish_order(workspace: &Path) -> Result<Vec<Package>> {
    let output = require_success(
        run(
            workspace,
            "cargo",
            &["metadata", "--no-deps", "--format-version", "1"],
        )?,
        "cargo metadata",
    )?;
    let metadata = serde_json::from_slice(&output.stdout).context("parsing cargo metadata")?;
    publish_order(metadata)
}

fn publish_order(metadata: Metadata) -> Result<Vec<Package>> {
    let members = metadata
        .workspace_members
        .into_iter()
        .collect::<BTreeSet<_>>();
    let mut packages = metadata
        .packages
        .into_iter()
        .filter(|package| members.contains(&package.id))
        .filter(|package| {
            package
                .publish
                .as_ref()
                .is_none_or(|registries| !registries.is_empty())
        })
        .map(|package| (package.name.clone(), package))
        .collect::<BTreeMap<_, _>>();
    let names = packages.keys().cloned().collect::<BTreeSet<_>>();
    let mut dependents = names
        .iter()
        .map(|name| (name.clone(), BTreeSet::new()))
        .collect::<BTreeMap<_, _>>();
    let mut in_degree = names
        .iter()
        .map(|name| (name.clone(), 0_usize))
        .collect::<BTreeMap<_, _>>();

    for package in packages.values() {
        for dependency in &package.dependencies {
            if names.contains(&dependency.name) {
                dependents
                    .get_mut(&dependency.name)
                    .unwrap()
                    .insert(package.name.clone());
                *in_degree.get_mut(&package.name).unwrap() += 1;
            }
        }
    }

    let mut ready = in_degree
        .iter()
        .filter(|(_, degree)| **degree == 0)
        .map(|(name, _)| name.clone())
        .collect::<VecDeque<_>>();
    let mut order = Vec::with_capacity(packages.len());
    while let Some(name) = ready.pop_front() {
        order.push(packages.remove(&name).unwrap());
        for dependent in dependents.get(&name).unwrap() {
            let degree = in_degree.get_mut(dependent).unwrap();
            *degree -= 1;
            if *degree == 0 {
                let position = ready
                    .iter()
                    .position(|queued| queued > dependent)
                    .unwrap_or(ready.len());
                ready.insert(position, dependent.clone());
            }
        }
    }
    if !packages.is_empty() {
        bail!("cycle detected in workspace dependency graph");
    }
    Ok(order)
}

fn allow_remediation(dry_run: bool) -> bool {
    !dry_run
}

fn run_gates(workspace: &Path, dry_run: bool) -> Result<bool> {
    step("Quality gates");
    let fmt = run(workspace, "cargo", &["fmt", "--all", "--check"])?;
    if !fmt.status.success() {
        if !allow_remediation(dry_run) {
            fail("fmt check failed; dry-run will not modify files");
            return Ok(false);
        }
        warn("fmt check failed — auto-formatting");
        require_success(run(workspace, "cargo", &["fmt", "--all"])?, "cargo fmt")?;
    }

    let clippy = run(
        workspace,
        "cargo",
        &["clippy", "--all-targets", "--", "-D", "warnings"],
    )?;
    if !clippy.status.success() {
        if !allow_remediation(dry_run) {
            fail("Clippy failed; dry-run will not apply fixes");
            return Ok(false);
        }
        warn("Clippy warnings found — attempting auto-fix");
        let _ = run(
            workspace,
            "cargo",
            &[
                "clippy",
                "--fix",
                "--allow-dirty",
                "--allow-staged",
                "--all-targets",
                "--",
                "-D",
                "warnings",
            ],
        )?;
        if !run(
            workspace,
            "cargo",
            &["clippy", "--all-targets", "--", "-D", "warnings"],
        )?
        .status
        .success()
        {
            fail("Clippy still failing after auto-fix");
            return Ok(false);
        }
    }

    if !run(workspace, "cargo", &["nextest", "run"])?
        .status
        .success()
    {
        fail("Tests failed");
        return Ok(false);
    }
    ok("Quality gates passed");
    Ok(true)
}

fn output_text(output: &Output) -> String {
    format!(
        "{}{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    )
}

fn is_already_published(output: &str) -> bool {
    let output = output.to_ascii_lowercase();
    ["already exists", "crate version `", "already uploaded"]
        .iter()
        .any(|marker| output.contains(marker))
}

fn is_rate_limited(output: &str) -> bool {
    let output = output.to_ascii_lowercase();
    ["too many requests", "rate limit", "429"]
        .iter()
        .any(|marker| output.contains(marker))
}

fn retry_delay(attempt: u32, base_ms: Option<u64>) -> Duration {
    let factor = 1_u64
        .checked_shl(attempt.saturating_sub(1))
        .unwrap_or(u64::MAX);
    match base_ms {
        Some(base) => Duration::from_millis(base.saturating_mul(factor)),
        None => Duration::from_secs(5_u64.saturating_mul(factor)),
    }
}

fn publish_crate(package: &Package, dry_run: bool, max_retries: u32) -> Result<bool> {
    let cwd = package
        .manifest_path
        .parent()
        .context("package manifest has no parent")?;
    let args = if dry_run {
        vec!["publish", "--dry-run"]
    } else {
        vec!["publish"]
    };
    let retry_base_ms = std::env::var("GODMODE_RELEASE_RETRY_BASE_MS")
        .ok()
        .and_then(|value| value.parse().ok());
    for attempt in 1..=max_retries {
        info(format!(
            "Publishing {} v{} (attempt {attempt}/{max_retries})",
            package.name, package.version
        ));
        let output = run(cwd, "cargo", &args)?;
        let text = output_text(&output);
        if output.status.success() || is_already_published(&text) {
            return Ok(true);
        }
        if attempt == max_retries {
            fail(format!("{} publish failed:\n{text}", package.name));
            return Ok(false);
        }
        let wait = retry_delay(attempt, retry_base_ms);
        if is_rate_limited(&text) {
            warn(format!(
                "Rate limited; retrying in {:.3}s",
                wait.as_secs_f64()
            ));
        } else {
            fail(format!("{} publish failed:\n{text}", package.name));
            warn(format!("Retrying in {:.3}s", wait.as_secs_f64()));
        }
        thread::sleep(wait);
    }
    Ok(false)
}

fn load_state(path: &Path) -> Result<ReleaseState> {
    if !path.exists() {
        return Ok(ReleaseState::default());
    }
    let raw =
        std::fs::read_to_string(path).with_context(|| format!("reading {}", path.display()))?;
    serde_json::from_str(&raw).with_context(|| format!("parsing {}", path.display()))
}

fn save_state(path: &Path, state: &ReleaseState) -> Result<()> {
    std::fs::write(path, format!("{}\n", serde_json::to_string_pretty(state)?))
        .with_context(|| format!("writing {}", path.display()))
}

fn emit_report(
    order: &[Package],
    results: &BTreeMap<String, ResultStatus>,
    dry_run: bool,
    elapsed: Duration,
) -> bool {
    println!(
        "\n{}\n  {}RELEASE REPORT\n{}",
        "═".repeat(60),
        if dry_run { "DRY-RUN " } else { "" },
        "═".repeat(60)
    );
    for package in order {
        let (icon, label) = match results.get(&package.name) {
            Some(ResultStatus::Published) => ("✓", "published"),
            Some(ResultStatus::Skipped) => ("↩", "skipped"),
            Some(ResultStatus::Failed) => ("✗", "failed"),
            None => ("?", "?"),
        };
        println!(
            "  {icon}  {:<30} v{}  [{label}]",
            package.name, package.version
        );
    }
    let count = |status| results.values().filter(|value| **value == status).count();
    println!("{}", "─".repeat(60));
    println!("  Published : {}", count(ResultStatus::Published));
    println!("  Skipped   : {}", count(ResultStatus::Skipped));
    println!("  Failed    : {}", count(ResultStatus::Failed));
    println!("  Elapsed   : {:.1}s", elapsed.as_secs_f64());
    if dry_run {
        println!("  This was a DRY RUN — nothing was published or written.");
    }
    count(ResultStatus::Failed) == 0
}

fn main() -> Result<()> {
    let args = Args::parse();
    let workspace = args
        .workspace
        .canonicalize()
        .with_context(|| format!("resolving {}", args.workspace.display()))?;
    let state_path = workspace.join(".release-state.json");
    let started = Instant::now();
    step("Rust Workspace Release Orchestrator");
    info(format!("Workspace: {}", workspace.display()));
    let mut state = if args.resume {
        load_state(&state_path)?
    } else {
        ReleaseState::default()
    };
    let order = compute_publish_order(&workspace)?;
    for (index, package) in order.iter().enumerate() {
        println!("  {}. {} v{}", index + 1, package.name, package.version);
    }

    if !args.skip_gates && !state.gates_passed {
        if !run_gates(&workspace, args.dry_run)? {
            if !args.dry_run {
                save_state(&state_path, &state)?;
            }
            bail!("quality gates failed; fix issues and re-run with --resume");
        }
        state.gates_passed = true;
        if !args.dry_run {
            save_state(&state_path, &state)?;
        }
    }

    let max_retries = std::env::var("GODMODE_RELEASE_MAX_RETRIES")
        .ok()
        .and_then(|value| value.parse::<u32>().ok())
        .filter(|value| *value > 0)
        .unwrap_or(5);
    let mut results = BTreeMap::new();
    for package in &order {
        if state.published.contains(&package.name) {
            results.insert(package.name.clone(), ResultStatus::Skipped);
            continue;
        }
        if publish_crate(package, args.dry_run, max_retries)? {
            results.insert(package.name.clone(), ResultStatus::Published);
            if !args.dry_run {
                state.published.push(package.name.clone());
                save_state(&state_path, &state)?;
            }
        } else {
            results.insert(package.name.clone(), ResultStatus::Failed);
            if !args.dry_run {
                save_state(&state_path, &state)?;
            }
            emit_report(&order, &results, args.dry_run, started.elapsed());
            bail!("{} failed; re-run with --resume", package.name);
        }
    }
    let success = emit_report(&order, &results, args.dry_run, started.elapsed());
    if success && !args.dry_run && state_path.exists() {
        std::fs::remove_file(&state_path)?;
    }
    if !success {
        bail!("release failed");
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn package(name: &str, dependencies: &[&str], publish: Option<Vec<String>>) -> Package {
        Package {
            id: format!("{name} 0.1.0"),
            name: name.into(),
            version: "0.1.0".into(),
            manifest_path: PathBuf::from(format!("/{name}/Cargo.toml")),
            publish,
            dependencies: dependencies
                .iter()
                .map(|name| Dependency {
                    name: (*name).into(),
                })
                .collect(),
        }
    }

    #[test]
    fn dependencies_publish_first_and_private_crates_are_skipped() {
        let packages = vec![
            package("cli", &["core", "adapters"], None),
            package("private", &[], Some(vec![])),
            package("adapters", &["core"], None),
            package("core", &[], None),
        ];
        let workspace_members = packages.iter().map(|package| package.id.clone()).collect();
        let order = publish_order(Metadata {
            packages,
            workspace_members,
        })
        .unwrap();
        assert_eq!(
            order
                .iter()
                .map(|package| package.name.as_str())
                .collect::<Vec<_>>(),
            ["core", "adapters", "cli"]
        );
    }

    #[test]
    fn cycles_are_rejected() {
        let packages = vec![package("a", &["b"], None), package("b", &["a"], None)];
        let workspace_members = packages.iter().map(|package| package.id.clone()).collect();
        assert!(publish_order(Metadata {
            packages,
            workspace_members
        })
        .is_err());
    }

    #[test]
    fn output_classification_is_case_insensitive() {
        assert!(is_already_published(
            "Crate version `1.0.0` is already uploaded"
        ));
        assert!(is_rate_limited("HTTP 429 Too Many Requests"));
    }

    #[test]
    fn dry_run_disables_mutating_gate_remediation() {
        assert!(!allow_remediation(true));
        assert!(allow_remediation(false));
    }

    #[test]
    fn retry_delay_can_be_disabled_for_fake_cargo_integration_tests() {
        assert_eq!(retry_delay(3, Some(0)), Duration::ZERO);
        assert_eq!(retry_delay(3, Some(2)), Duration::from_millis(8));
    }

    #[test]
    fn state_roundtrips() {
        let state = ReleaseState {
            published: vec!["core".into()],
            gates_passed: true,
        };
        assert_eq!(
            serde_json::from_str::<ReleaseState>(&serde_json::to_string(&state).unwrap()).unwrap(),
            state
        );
    }
}
